common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java

```diff
@@ -198,181 +177,463 @@
+                    holder = extensionInstances.get(url);
+                }
+                Object instance = holder.get();
+                if (instance == null) {
+                    synchronized (holder) {
+                        instance = holder.get();
+                        if (instance == null) {
+                            instance = createExtension(url, loader, argTypes, args);
+                            holder.set(instance);
+                        }
+                    }
+                }
+                return (S)instance;
             } else {
-                throw new EnhancedServiceNotFoundException(
-                    "not found service provider for : " + service.getName() + " caused by " + ExceptionUtils
-                        .getFullStackTrace(e));
+                return createNewExtension(url, loader, argTypes, args);
             }
         }
-    }
 
-    @SuppressWarnings("rawtypes")
-    private static <S> List<Class> findAllExtensionClass(Class<S> service, String activateName, ClassLoader loader) {
-        List<Class> extensions = new ArrayList<>();
-        try {
-            loadFile(service, SERVICES_DIRECTORY, loader, extensions);
-            loadFile(service, SEATA_DIRECTORY, loader, extensions);
-        } catch (IOException e) {
-            throw new EnhancedServiceNotFoundException(e);
+        private S createExtension(ExtensionURL url, ClassLoader loader, Class[] argTypes,
+                                  Object[] args) {
+            Class<?> clazz = loadAllExtensionClass(loader).get(url);
+            if (clazz == null) {
+                throw new EnhancedServiceNotFoundException("Not Found");
+            }
+            try {
+                S instance = (S)EXTENSION_INSTANCES.get(clazz);
+                if (instance == null) {
+                    S newInstance = initInstance(clazz, argTypes, args);
+                    EXTENSION_INSTANCES.putIfAbsent(clazz, newInstance);
+                    instance = newInstance;
+                }
+                return instance;
+            } catch (Throwable t) {
+                throw new IllegalStateException("Extension instance(url: " + url + ", class: " +
+                        type + ")  could not be instantiated: " + t.getMessage(), t);
+            }
         }
 
-        if (extensions.isEmpty()) {
-            return extensions;
-        }
-        extensions.sort((c1, c2) -> {
-            int o1 = 0;
-            int o2 = 0;
-            @SuppressWarnings("unchecked")
-            LoadLevel a1 = (LoadLevel) c1.getAnnotation(LoadLevel.class);
-            @SuppressWarnings("unchecked")
-            LoadLevel a2 = (LoadLevel) c2.getAnnotation(LoadLevel.class);
-
-            if (a1 != null) {
-                o1 = a1.order();
+        private S createNewExtension(ExtensionURL url, ClassLoader loader, Class[] argTypes, Object[] args) {
+            Class<?> clazz = loadAllExtensionClass(loader).get(url);
```

Comment:
If ExtensionURL has the service-class as field, calling loadAllExtensionClass is not necessary.

Reply:
This part of the code which use Holder and Double-Checkd Locking refers to Dubbo SPI. I think it has two functions at present:
1.When concurrent, ensure that an extensionDefinition use the same lock object
2.Whether the value in the Holder is null can be used as the judgment condition of Double-Checked Locking
Without using Holder here,i think it is difficult to guarantee the above two points at the same time.

Reply:
Yeah,good idea,i add serviceClass field and change the data structure of  Map（ExtensionDefinition_To_Class） to List<ExtensionDefintion>